In [58]:
import pandas as pd

In [59]:
know_2018 = pd.read_csv('../know_Data/train/KNOW_2018 밀린부분 복원.csv', encoding='cp949')

C:\Users\user\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3444: DtypeWarning: Columns (136) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [60]:
know_2018

idx  cq1  cq2  cq3  cq4  cq5  cq6  cq7  cq8  cq9  ...  bq37_1  bq38  \
0      9486    5    3    4    2    2    5    4    5    4  ...    치기공과     1   
1      9487    4    3    4    4    4    4    4    4    4  ...      이과     1   
2      9488    4    1    5    3    3    4    2    4    1  ...     기계과     2   
3      9489    4    3    4    4    4    3    3    2    3  ...   물리 과학     1   
4      9490    3    2    3    1    1    4    3    4    3  ...    치기공학     1   
...     ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...     ...   ...   
9067  18564    5    5    5    2    2    4    2    5    3  ...     한약학     2   
9068  18565    5    5    5    3    3    5    4    4    4  ...     국문과     2   
9069  18566    4    4    4    3    3    5    4    4    4  ...      의대     2   
9070  18567    5    2    3    2    3    5    4    4    3  ...      의예     2   
9071  18568    5    4    5    4    3    5    4    1    4  ...   컴퓨터공학     1   

      bq38_1  bq38_2  bq39  bq40  bq41_1  bq41_2  bq41_3  knowcode  
0          1             1    42    3800    3400            306301  
1          2             1    40    2300    1800            133204  
2                  6          30                    3000    702502  
3          1             1    40    3350    2750            121201  
4          1             1    40    2500    2300            306301  
...      ...     ...   ...   ...     ...     ...     ...       ...  
9067               6          60                    2000    303002  
9068               6          52                    5000    615203  
9069               5          38                    9000    301106  
9070               5          56                    6500    301112  
9071       1             1    40    4200    2800            133101  

[9072 rows x 141 columns]

In [61]:
know_2018.isnull().sum()[:50]

idx     0
cq1     0
cq2     0
cq3     0
cq4     0
cq5     0
cq6     0
cq7     0
cq8     0
cq9     0
cq10    0
cq11    0
cq12    0
cq13    0
cq14    0
cq15    0
cq16    0
cq17    0
cq18    0
cq19    0
cq20    0
cq21    0
cq22    0
cq23    0
cq24    0
cq25    0
cq26    0
cq27    0
cq28    0
cq29    0
cq30    0
cq31    0
cq32    0
cq33    0
cq34    0
cq35    0
cq36    0
cq37    0
cq38    0
cq39    0
cq40    0
cq41    0
cq42    0
cq43    0
cq44    0
cq45    0
cq46    0
cq47    0
cq48    0
cq49    0
dtype: int64

In [62]:
know_2018.isnull().sum()[50:100]

cq50_1    0
cq50_2    0
cq50_3    0
cq50_4    0
cq50_5    0
cq50_6    0
cq50_7    0
cq50_8    0
iq1       0
iq2       0
iq3       0
iq4       0
iq5       0
iq6       0
bq1       0
bq2       0
bq3       0
bq4       0
bq4_1a    0
bq4_1b    0
bq4_1c    0
bq5       0
bq5_1     0
bq5_2     0
bq6       0
bq7       0
bq8_1     0
bq8_2     0
bq8_3     0
bq9       0
bq10      0
bq11      0
bq12_1    0
bq12_2    0
bq12_3    0
bq12_4    0
bq12_5    0
bq13      0
bq14      0
bq15      0
bq16      0
bq17      0
bq18      0
bq19      0
bq20      0
bq21      0
bq221     0
bq222     0
bq223     0
bq231     0
dtype: int64

In [63]:
know_2018.isnull().sum()[100:]

bq232        0
bq233        0
bq234        0
bq235        0
bq241        0
bq242        0
bq243        0
bq244        0
bq245        0
bq25         0
bq25_1       0
bq26_1       0
bq26_1a      0
bq26_2       0
bq26_2a      0
bq26_3       0
bq26_3a      0
bq26_4       0
bq26_4a      0
bq27         0
bq28         0
bq28_1       0
bq29         0
bq30         0
bq31         0
bq32         1
bq33         1
bq34         0
bq35         0
bq36         0
bq37         0
bq37_1       0
bq38         0
bq38_1       1
bq38_2       0
bq39         1
bq40         0
bq41_1       1
bq41_2       1
bq41_3      32
knowcode     0
dtype: int64

In [77]:
pd.set_option('display.max_columns' ,None)
pd.set_option('display.max_rows' ,50)

In [78]:
know_2018[know_2018.bq38_1.isnull()]


idx  cq1  cq2  cq3  cq4  cq5  cq6  cq7  cq8  cq9  cq10  cq11  cq12  \
8551  18048    5    4    3    3    1    5    2    2    1     1     2     2   

      cq13  cq14  cq15  cq16  cq17  cq18  cq19  cq20  cq21  cq22  cq23  cq24  \
8551     2     5     1     1     3     5     1     1     2     1     1     1   

      cq25  cq26  cq27  cq28  cq29  cq30  cq31  cq32  cq33  cq34  cq35  cq36  \
8551     1     1     1     1     5     1     1     5     1     5     1     2   

      cq37  cq38  cq39  cq40  cq41  cq42  cq43  cq44  cq45  cq46  cq47  cq48  \
8551     3     4     3     1     4     5     3     1     4     5     1     1   

      cq49  cq50_1  cq50_2  cq50_3  cq50_4  cq50_5  cq50_6  cq50_7  cq50_8  \
8551     2       1       3       1       1       1       1       1       1   

      iq1  iq2  iq3  iq4  iq5  iq6  bq1  bq2  bq3  bq4 bq4_1a     bq4_1b  \
8551    1    4    5    1    2    1   13    1    1    1  컬러리스트  섬유디자인산업기사   

     bq4_1c  bq5 bq5_1         bq5_2  bq6  bq7  bq8_1  bq8_2  bq8_3  bq9  \
8551           1     3  배색 및 컬러매칭 훈련    5    5      5      5      5    4   

      bq10  bq11  bq12_1 bq12_2 bq12_3 bq12_4  bq12_5  bq13  bq14  bq15  bq16  \
8551     5     3       5      5      5      5       5     3     2     1     2   

      bq17  bq18  bq19  bq20  bq21 bq221 bq222  bq223 bq231  bq232  bq233  \
8551     1     1     1     3   100     1     2      0     1      0      0   

      bq234 bq235 bq241 bq242 bq243 bq244 bq245  bq25 bq25_1  bq26_1 bq26_1a  \
8551      0           1                             1      2       1       2   

      bq26_2 bq26_2a  bq26_3 bq26_3a  bq26_4 bq26_4a  bq27  bq28  \
8551       1       2       1       2       1       2     2     3   

                                                 bq28_1 bq29  \
8551  포지션이 희소성이 높고 중요도 또한 높아 원래대로라면 늘어나야 하는 것이 맞지만_경...  배색사   

                   bq30 bq31 bq32  \
8551  포토샵, 일러스트레이터 프로그램   없음   없음   

                                                   bq33  bq34  bq35  bq36  \
8551  컬러리스트는 분야별로 종류가 다양하며 제가 종사하고 있는 분야는 패션입니다._패션 ...     2     2    37   

      bq37 bq37_1  bq38 bq38_1 bq38_2 bq39 bq40 bq41_1 bq41_2 bq41_3  knowcode  
8551     5    색채학     2    NaN      5  NaN   50    NaN    NaN   5000    415404

In [79]:
know_2018[know_2018.bq39.isnull()]

idx  cq1  cq2  cq3  cq4  cq5  cq6  cq7  cq8  cq9  cq10  cq11  cq12  \
8551  18048    5    4    3    3    1    5    2    2    1     1     2     2   

      cq13  cq14  cq15  cq16  cq17  cq18  cq19  cq20  cq21  cq22  cq23  cq24  \
8551     2     5     1     1     3     5     1     1     2     1     1     1   

      cq25  cq26  cq27  cq28  cq29  cq30  cq31  cq32  cq33  cq34  cq35  cq36  \
8551     1     1     1     1     5     1     1     5     1     5     1     2   

      cq37  cq38  cq39  cq40  cq41  cq42  cq43  cq44  cq45  cq46  cq47  cq48  \
8551     3     4     3     1     4     5     3     1     4     5     1     1   

      cq49  cq50_1  cq50_2  cq50_3  cq50_4  cq50_5  cq50_6  cq50_7  cq50_8  \
8551     2       1       3       1       1       1       1       1       1   

      iq1  iq2  iq3  iq4  iq5  iq6  bq1  bq2  bq3  bq4 bq4_1a     bq4_1b  \
8551    1    4    5    1    2    1   13    1    1    1  컬러리스트  섬유디자인산업기사   

     bq4_1c  bq5 bq5_1         bq5_2  bq6  bq7  bq8_1  bq8_2  bq8_3  bq9  \
8551           1     3  배색 및 컬러매칭 훈련    5    5      5      5      5    4   

      bq10  bq11  bq12_1 bq12_2 bq12_3 bq12_4  bq12_5  bq13  bq14  bq15  bq16  \
8551     5     3       5      5      5      5       5     3     2     1     2   

      bq17  bq18  bq19  bq20  bq21 bq221 bq222  bq223 bq231  bq232  bq233  \
8551     1     1     1     3   100     1     2      0     1      0      0   

      bq234 bq235 bq241 bq242 bq243 bq244 bq245  bq25 bq25_1  bq26_1 bq26_1a  \
8551      0           1                             1      2       1       2   

      bq26_2 bq26_2a  bq26_3 bq26_3a  bq26_4 bq26_4a  bq27  bq28  \
8551       1       2       1       2       1       2     2     3   

                                                 bq28_1 bq29  \
8551  포지션이 희소성이 높고 중요도 또한 높아 원래대로라면 늘어나야 하는 것이 맞지만_경...  배색사   

                   bq30 bq31 bq32  \
8551  포토샵, 일러스트레이터 프로그램   없음   없음   

                                                   bq33  bq34  bq35  bq36  \
8551  컬러리스트는 분야별로 종류가 다양하며 제가 종사하고 있는 분야는 패션입니다._패션 ...     2     2    37   

      bq37 bq37_1  bq38 bq38_1 bq38_2 bq39 bq40 bq41_1 bq41_2 bq41_3  knowcode  
8551     5    색채학     2    NaN      5  NaN   50    NaN    NaN   5000    415404

In [64]:
know_2018.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9072 entries, 0 to 9071
Data columns (total 141 columns):
 #    Column    Dtype 
---   ------    ----- 
 0    idx       int64 
 1    cq1       int64 
 2    cq2       int64 
 3    cq3       int64 
 4    cq4       int64 
 5    cq5       int64 
 6    cq6       int64 
 7    cq7       int64 
 8    cq8       int64 
 9    cq9       int64 
 10   cq10      int64 
 11   cq11      int64 
 12   cq12      int64 
 13   cq13      int64 
 14   cq14      int64 
 15   cq15      int64 
 16   cq16      int64 
 17   cq17      int64 
 18   cq18      int64 
 19   cq19      int64 
 20   cq20      int64 
 21   cq21      int64 
 22   cq22      int64 
 23   cq23      int64 
 24   cq24      int64 
 25   cq25      int64 
 26   cq26      int64 
 27   cq27      int64 
 28   cq28      int64 
 29   cq29      int64 
 30   cq30      int64 
 31   cq31      int64 
 32   cq32      int64 
 33   cq33      int64 
 34   cq34      int64 
 35   cq35      int64 
 36   cq36      in

In [65]:
for i in know_2018.columns : 
    print(i, len(know_2018[i].value_counts()))

idx 9072
cq1 5
cq2 5
cq3 5
cq4 5
cq5 5
cq6 5
cq7 5
cq8 5
cq9 5
cq10 5
cq11 5
cq12 5
cq13 5
cq14 5
cq15 5
cq16 5
cq17 5
cq18 5
cq19 5
cq20 5
cq21 5
cq22 5
cq23 5
cq24 5
cq25 5
cq26 5
cq27 5
cq28 5
cq29 5
cq30 5
cq31 5
cq32 5
cq33 5
cq34 5
cq35 5
cq36 5
cq37 5
cq38 5
cq39 5
cq40 5
cq41 5
cq42 5
cq43 5
cq44 5
cq45 5
cq46 5
cq47 3
cq48 2
cq49 5
cq50_1 3
cq50_2 3
cq50_3 3
cq50_4 3
cq50_5 3
cq50_6 3
cq50_7 3
cq50_8 3
iq1 5
iq2 5
iq3 5
iq4 5
iq5 5
iq6 5
bq1 21
bq2 6
bq3 6
bq4 2
bq4_1a 1596
bq4_1b 783
bq4_1c 320
bq5 2
bq5_1 7
bq5_2 3804
bq6 7
bq7 7
bq8_1 5
bq8_2 5
bq8_3 5
bq9 5
bq10 5
bq11 5
bq12_1 5
bq12_2 6
bq12_3 6
bq12_4 6
bq12_5 5
bq13 3
bq14 2
bq15 4
bq16 2
bq17 3
bq18 3
bq19 3
bq20 5
bq21 44
bq221 4
bq222 2
bq223 2
bq231 6
bq232 4
bq233 2
bq234 2
bq235 2
bq241 5
bq242 3
bq243 2
bq244 2
bq245 2
bq25 3
bq25_1 3
bq26_1 3
bq26_1a 4
bq26_2 3
bq26_2a 4
bq26_3 3
bq26_3a 4
bq26_4 3
bq26_4a 4
bq27 2
bq28 5
bq28_1 7671
bq29 1652
bq30 5859
bq31 1834
bq32 2314
bq33 705
bq34 5
bq35 2
bq36 67
bq37 6


1. 결측치

- 주관식 
bq32         1
bq33         1

- 처리해야하는 col
bq38_1       1
bq39         1
bq41_1       1
bq41_2       1
bq41_3      32

2. 이상치 

3. 주관식

- 자격증
bq4_1a 1596
bq4_1b 783
bq4_1c 320

- 버림
bq5_2 3804

bq28_1 7671
bq29 1652
bq30 5859
bq31 1834
bq32 2314
bq33 705

- 학과 
bq37_1 1825

4. 숫자여야하는 col

bq21 44
bq36 67
bq40 132
bq41_1 273
bq41_2 182
bq41_3 94

5. 체크해봐야하는 col

bq221 4
bq222 2
bq223 2
bq231 6
bq232 4
bq233 2
bq234 2
bq235 2
bq241 5
bq242 3
bq243 2
bq244 2
bq245 2

## 이상치 처리

In [17]:
for col in know_2018:
        know_2018[col].replace(' ', 0, inplace = True)

In [22]:
str_li =['bq4_1a','bq4_1b','bq4_1c','bq5_2','bq28_1','bq29','bq30','bq31', 'bq32','bq33','bq37_1']


In [23]:
know_2018 = know_2018.fillna(0)

In [24]:
know_2018.isnull().sum()[100:]

bq232       0
bq233       0
bq234       0
bq235       0
bq241       0
bq242       0
bq243       0
bq244       0
bq245       0
bq25        0
bq25_1      0
bq26_1      0
bq26_1a     0
bq26_2      0
bq26_2a     0
bq26_3      0
bq26_3a     0
bq26_4      0
bq26_4a     0
bq27        0
bq28        0
bq28_1      0
bq29        0
bq30        0
bq31        0
bq32        0
bq33        0
bq34        0
bq35        0
bq36        0
bq37        0
bq37_1      0
bq38        0
bq38_1      0
bq38_2      0
bq39        0
bq40        0
bq41_1      0
bq41_2      0
bq41_3      0
knowcode    0
dtype: int64

In [25]:
for i in know_2018.columns : 
    if i not in str_li : 
        know_2018[i] = know_2018[i].astype(int)
    else :
        know_2018[i]=know_2018[i].astype(str)


In [26]:
know_2018.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9072 entries, 0 to 9071
Data columns (total 141 columns):
 #    Column    Dtype 
---   ------    ----- 
 0    idx       int32 
 1    cq1       int32 
 2    cq2       int32 
 3    cq3       int32 
 4    cq4       int32 
 5    cq5       int32 
 6    cq6       int32 
 7    cq7       int32 
 8    cq8       int32 
 9    cq9       int32 
 10   cq10      int32 
 11   cq11      int32 
 12   cq12      int32 
 13   cq13      int32 
 14   cq14      int32 
 15   cq15      int32 
 16   cq16      int32 
 17   cq17      int32 
 18   cq18      int32 
 19   cq19      int32 
 20   cq20      int32 
 21   cq21      int32 
 22   cq22      int32 
 23   cq23      int32 
 24   cq24      int32 
 25   cq25      int32 
 26   cq26      int32 
 27   cq27      int32 
 28   cq28      int32 
 29   cq29      int32 
 30   cq30      int32 
 31   cq31      int32 
 32   cq32      int32 
 33   cq33      int32 
 34   cq34      int32 
 35   cq35      int32 
 36   cq36      in

In [27]:
# bq37=1(중졸)인 사람중 학과 입력한 사람
know_2018[(know_2018.bq37==1) & (know_2018.bq37_1!='0')]['bq37_1'].value_counts()

일반    1
Name: bq37_1, dtype: int64

In [29]:
know_2018[(know_2018.bq37==1) & (know_2018.bq37_1!='0')].index

Int64Index([7589], dtype='int64')

In [31]:
know_2018.iloc[7589,131]='0'

In [34]:
# 고졸 이상중 학과 입력안한 사람 
# 앞에서 '' 을 0 으로 replace했기 때문에 0으로 찾아야함 
know_2018[(know_2018.bq37.isin([2,3,4,5,6])) & (know_2018.bq37_1=='0')]['bq37'].value_counts()

2    131
4     32
3     17
5      4
6      1
Name: bq37, dtype: int64

In [35]:
# 근로시간
know_2018[know_2018.bq40==0].index

Int64Index([2590], dtype='int64')

In [37]:
know_2018.iloc[2590,136] =42

In [39]:
# 임금근로자

null_li = know_2018[(know_2018.bq38==1) & (know_2018.bq41_1==0)].index
null_li

Int64Index([85, 90, 120, 141, 1297, 3486, 5735, 6153], dtype='int64')

In [45]:
for i in null_li:     
    know_2018.iloc[i,137]=round(know_2018[(know_2018.bq1 ==know_2018.iloc[i,64])&(know_2018.bq41_1!=0)].bq41_1.median())

In [46]:
null_li = know_2018[(know_2018.bq38==1) & (know_2018.bq41_1==0)].index
null_li

Int64Index([], dtype='int64')

In [47]:
# 임금근로자 초봉
null_li2 = know_2018[(know_2018.bq38==1) & (know_2018.bq41_2==0)].index
null_li2

Int64Index([  28,   85,   90,  115,  120,  141,  172,  426,  496,  538,  658,
             811, 1146, 1180, 1209, 1297, 1353, 1418, 1474, 1490, 1519, 1601,
            1630, 1667, 1740, 1746, 1855, 1859, 1936, 2181, 2335, 2420, 2458,
            2776, 3061, 3158, 3201, 3249, 3276, 3313, 3335, 3388, 3474, 3486,
            3645, 3694, 3863, 3899, 3925, 4162, 4236, 4550, 4655, 4775, 4817,
            5008, 5055, 5299, 5333, 5442, 5545, 5612, 5694, 5735, 5890, 5982,
            6153, 6165, 6307, 6337, 6344, 6522, 6596, 6732, 6758, 6768, 6878,
            7106, 7135, 7152, 7259, 7286, 7439, 7470, 7536, 7738, 7855, 8133,
            8188, 8415, 8467, 8576, 8731, 8935],
           dtype='int64')

In [48]:
for i in null_li2:     
    know_2018.iloc[i,138]=round(know_2018[(know_2018.bq1 ==know_2018.iloc[i,64])&(know_2018.bq41_2!=0)].bq41_2.median())

In [49]:
null_li2 = know_2018[(know_2018.bq38==1) & (know_2018.bq41_2==0)].index
null_li2

Int64Index([], dtype='int64')

In [50]:
# 비임금근로자 근로소득

null_li3 = know_2018[(know_2018.bq38==2) & (know_2018.bq41_3==0)].index
null_li3

Int64Index([ 183,  458,  919, 1028, 1728, 1875, 2115, 3475, 3644, 3814, 4760,
            6066, 7557, 7929, 8371, 8428, 8534, 8560],
           dtype='int64')

In [51]:
for i in null_li3:     
    know_2018.iloc[i,139]=round(know_2018[(know_2018.bq1 ==know_2018.iloc[i,64])&(know_2018.bq41_3!=0)].bq41_3.median())

In [52]:
null_li3 = know_2018[(know_2018.bq38==2) & (know_2018.bq41_3==0)].index
null_li3

Int64Index([], dtype='int64')

In [54]:
## 자격증 없음

know_2018['bq4_1b'].replace(0, '0', inplace = True)


for i in know_2018.bq4_1b:
    if '없' in i : 
        print(i)

# know_2020['bq4_1b'].replace('없다', '0', inplace = True)

In [55]:
know_2018['bq4_1c'].replace(0, '0', inplace = True)


for i in know_2018.bq4_1c:
    if '없' in i : 
        print(i)

# know_2020['bq4_1b'].replace('없다', '0', inplace = True)